Server/MirEnvir/Envir.cs

```diff
@@ -2532,7 +2533,93 @@ public bool CharacterExists(string name)
 
             return false;
         }
+        public List<CharacterInfo> MatchPlayer(string PlayerID, bool match = false)
+        {
+            if (string.IsNullOrEmpty(PlayerID)) return new List<CharacterInfo>(CharacterList);
+
+            List<CharacterInfo> list = new List<CharacterInfo>();
+
+            for (int i = 0; i < CharacterList.Count; i++)
+            {
+                if (match)
+                {
+                    if (CharacterList[i].Name.Equals(PlayerID, StringComparison.OrdinalIgnoreCase))
+                        list.Add(CharacterList[i]);
+                }
+                else
+                {
+                    if (CharacterList[i].Name.IndexOf(PlayerID, StringComparison.OrdinalIgnoreCase) >= 0)
+                        list.Add(CharacterList[i]);
+                }
+            }
+
+            return list;
+        }
+        public List<CharacterInfo> MatchPlayerbyItem(string ItemID, bool match = false)
+        {
+            if (string.IsNullOrEmpty(ItemID)) return new List<CharacterInfo>(CharacterList);
+
+            ulong itemId = 0;
+            ulong.TryParse(ItemID, out itemId);
+
+            List<CharacterInfo> list = new List<CharacterInfo>();
+
+            for (int i = 0; i < CharacterList.Count; i++)
```

Comment:
i dont get the difference between match and not a match?

Reply:
When match is set to false, the method does not check:

AccountInfo.Storage: Excludes checking items stored in the account storage.
Mail: Excludes checking items attached to mails.

Reply:
In that case just wrap the check around those loops which the "match" utilises, so you don't have to duplicate the entire code

Reply:
I have no clue how

Reply:
get rid of the if(match) else and and wrap the foreach loops which you only want to be used when if(match) is true

Reply:
Fixed